### ISSUE TITLE
WebSock broadcast vs commands

### ISSUE THREAD

**Issue by pedromorgan** [ts2/ts2-sim-server#13]
Raising issue from PR https://github.com/ts2/ts2-sim-server/pull/11

debate is that as usual devs have different expectations...and what it does essentially..
so here's pedro views..

**Comment by pedromorgan** [comment]
# Broadcast
- expects a general state of the sim/clock etc to be broadcast regular
- even a "tick" pulse with ts, simSpeed, sim.clock, virtual.time, running etc eg

To connect up to other devices, then a "tick" needs to be send, probably every 1100.ms just to freak some people out ocassinally ;-)

**Comment by pedromorgan** [comment]
# Interlocking
- the mad idea  in @pedromorgan  head is still to connect to a train set/layout
-  devices, and make it as real as possible..
- so keep in mind that imputs to sim, might be not a sim but REAL!!

**Comment by pedromorgan** [comment]
https://github.com/ts2/ts2-sim-server/pull/11#issuecomment-489341376
```
Responses are sent only to the client that sent the command. Other clients should register to listeners to be notified. Here the "stateChanged" event is what you need.
```

The snag with above, is that one needs a connection, and then addListenerat and dies...
This means that every client will have to add a listener, which is what we want..
Am attempting atmo to connect as many devices to the sim, and make them all maintain state.. eg
- Control panel on mobile
- a Tablet computer pretending to be arrival/departures board
- my pc running it all ;-)
- Some part of me still thinks we can manipulate in svg, dont know why.. but for some reason svg

**Comment by npiganeau** [comment]
Websocket API won't change in the short term, because other people need it frozen and it does the job (I managed to port the python client).

Broadcast: it does it as long as you register to the relevent events. See events.go for available events. Maybe some are missing for service change or train type change.

Interlocking: that's the mid-term target. Relevent code is out in the plugin packages so it can be modified to connect to real stuff.

Multiple clients as described is possible already and is my target too.

**Comment by pedromorgan** [comment]
Multiple clients is my main goal, remember the debate we had about ws yonks ago ;-))))

**Comment by pedromorgan** [comment]
# Commands

assuming many clients listening.. and everything red is that
1 - u are registered as a client
2 - u send the sim stop command
3 - restart etc and send command to all clients

So the command does have to propogate across all the clients ??

`

**Comment by npiganeau** [comment]
Clients don't keep state. They get notified when something changes and update accordingly.

**Comment by pedromorgan** [comment]
Only if they have listenerers added since last connection ...

# Editor

So see u playing there,
big question is how do we migrate olde sims into new ??
is that something we need to do right ??

I also assume that the editor in py, knocks outs the "sim.json".. ?
and then restart the server is the  idea ??

Bring me up to speed @npiganeau

**Comment by npiganeau** [comment]
You have server.renotify for that issue. It will send back last notifications that you might have lost. See also python client implementation. Almost everything depends on events and renotify does the trick at startup.

Editor: yes this is the idea and my work of the day. Should be finished in a couple of days at most.

**Comment by pedromorgan** [comment]
Ok so this works here, main change is that request is handed over to repsonse for object,action+optional.ID

**Comment by pedromorgan** [comment]
Quick question @npiganeau , as am reasearching..
is there a list of events..
I want to addlistener to sim.start and sim.stop...

**Comment by npiganeau** [comment]
See above: events.go for the list, 'stateChanged' is the event you need

**Comment by pedromorgan** [comment]
Yea i see the events, but how do i add.. eg Send >>
```{"object": "server", "action": "addListener", "params": {"event": "simulation", "ids": []}}``

Sorry to be a pain, am just trying to figure it out.

**Comment by npiganeau** [comment]
```{"object": "server", "action": "addListener", "params": {"event": "stateChanged"}}```

Then you start or pause the simulation and you will have the notification.

**Comment by pedromorgan** [comment]
don't seem to do anything..
```
> SENT: {"object": "server", "action": "register", "params": {"type": "client", "token": "client-secret"}}
= RESPONSE: {"id":0,"msgType":"response","data":{"status":"OK","message":"Successfully registered"}}

> SENT: {"object": "server", "action": "addListener", "params": {"event": "stateChanged"}}
= RESPONSE: {"id":0,"msgType":"response","data":{"status":"OK","message":"Listener added successfully"}}

> SENT: {"object": "simulation", "action": "start"}
= RESPONSE: {"id":0,"msgType":"response","data":{"status":"OK","message":"Simulation started successfully"}}

> SENT: {"object": "simulation", "action": "pause"}
= RESPONSE: {"id":0,"msgType":"response","data":{"status":"OK","message":"Simulation paused successfully"}}

> SENT: {"object": "simulation", "action": "start"}
= RESPONSE: {"id":0,"msgType":"response","data":{"status":"OK","message":"Simulation started successfully"}}
```

**Comment by pedromorgan** [comment]
Apologies, the above does work (wrong js code)...
I get 
```{"msgType":"notification","data":{"name":"stateChanged","object":{"value":false}}}```

**Comment by npiganeau** [comment]
Yes, that's it. If value is true the sim is now running, if it is false, the sim is now paused. Every client that subscribed get notified. Check with the pyclient.

Other events send other objects, depending on the event. E.g. routeActivated returns the route that has just been activated, trackItemChanged returns the track item that changed, etc.
